libminifi/include/utils/ThreadPool.h

```diff
@@ -303,8 +303,9 @@ class ThreadPool {
    * Drain will notify tasks to stop following notification
    */
   void drain() {
+    worker_queue_.stop();
     while (current_workers_ > 0) {
-      tasks_available_.notify_one();
+      std::this_thread::sleep_for(std::chrono::milliseconds(1));
```

Comment:
What is the purpose of this sleep?

Reply:
In my understanding, it is to wait for all worker thread functions to return. They are either in a task or waiting for one. If they are waiting for one, worker_queue_.stop() will make the tryDeqeue return with false and they will end. If they are in a task, they will finish executing it, call tryDeqeue, which will return false, and they will end.

My initial reaction at first was "why are we not using the ConcurrentQueue's cv instead of "polling" but I realized that is more complex than that.

Reply:
@bakaid is right. 
Stopping the queue wakes up all the worker threads actually doing nothing (waiting for work), but there can be worker threads doing some useful work (ontrigger calls, c2 heartbeats, whatever). These should end in a timely manner, but I found no better option to wait for that.

Reply:
Could you add a code comment explaining this, including the fact that this is a best-effort solution? It will be helpful to future readers of the code.

Reply:
Added